### PR TITLE
[SPARK-44030][SQL] Implement DataTypeExpression to offer Unapply for expression 

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Implement DataTypeExpression to offer `Unapply` for expression. By doing so we can drop `Unapply` from DataType.

### Why are the changes needed?

Simplify DataType interface.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala [hunk 1]
```diff
@@ -16,14 +16,36 @@
...
+  /**
+   * Enables matching against DataType for expressions:
+   * {{{
+   *   case Cast(child @ BinaryType(), StringType) =>
+   *     ...
+   * }}}
+   *
+   * @since 3.5.0
+   */
+  private[sql] def unapply(dataType: DataType, e: Expression): Boolean = e.dataType == dataType
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala:44]: If we can I would just remove the extractor.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala [hunk 2]
```diff
@@ -16,14 +16,36 @@
...
-import org.apache.spark.sql.catalyst.expressions.Cast
+import org.apache.spark.sql.catalyst.expressions.{Cast, Expression}
+import org.apache.spark.sql.catalyst.util.DataTypeJsonUtils.{DataTypeJsonDeserializer, DataTypeJsonSerializer}
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.internal.SQLConf.StoreAssignmentPolicy
 import org.apache.spark.sql.internal.SQLConf.StoreAssignmentPolicy.{ANSI, STRICT}
 import org.apache.spark.sql.types.{ArrayType, AtomicType, DataType, MapType, NullType, StructType}
 
+@JsonSerialize(using = classOf[DataTypeJsonSerializer])
+@JsonDeserialize(using = classOf[DataTypeJsonDeserializer])
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala:33]: I'm a bit confused, why we need to move these two annotations here? Need to use Jackson to serialize/deserialize object `DataTypeUtils`?

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala:33]: I suggest moving these two annotations back to DataType and we can change the following 

https://github.com/apache/spark/blob/daee47a436473b865343301a78fef715d2780c4d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DataTypeJsonUtils.scala#L37

to `delegate.serialize(DataTypeUtils.jsonValue(value.typeName), gen, provider)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/DataTypeUtils.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+object NumericTypeExpression {
+  /**
+   * Enables matching against NumericType for expressions:
+   * {{{
+   *   case Cast(child @ NumericType(), StringType) =>
+   *     ...
+   * }}}
+   */
+  def unapply(e: Expression): Boolean = {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala:58]: Why doesn't `e.dataType.isInstanceOf[NumericType]` work? The type hierarchy still exists right?

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala:58]: Well after a second thought, I think you are right that the implementation can be as easy as a few lines. Though all the tests have passed by the current way.

I changed the implementation to `e.dataType.isInstanceOf[NumericType]` and same for `IntegralType`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+   *     ...
+   * }}}
+   */
+  private[sql] def unapply(e: Expression): Boolean = e.dataType == dataType
+}
+
+abstract class AtomicTypeExpression(override val dataType: DataType)
+  extends DataTypeExpression(dataType)
+
+abstract class DatetimeTypeExpression(override val dataType: DataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala:35]: Seems the above two abstract classes are not needed.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala (hunk 4 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala:35]: oh yes good point!

**Review by hvanhovell — APPROVED** [review]
LGTM - merging.

**Pull request closed without merging**